jobs/testworker.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package jobs
+
+import (
+	"context"
+	"time"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/store"
+)
+
+type TestWorker struct {
+	store store.Store
+
+	name    string
+	stop    chan bool
+	stopped chan bool
+	jobs    model.JobChannel
+}
+
+func MakeTestWorker(s store.Store, name string) *TestWorker {
+	return &TestWorker{
+		store:   s,
+		name:    name,
+		stop:    make(chan bool, 1),
+		stopped: make(chan bool, 1),
+		jobs:    make(model.JobChannel),
+	}
+}
+
+func (worker *TestWorker) Run() {
+	l4g.Debug("Worker %v: Started", worker.name)
+
+	running := true
+	for running {
+		select {
+		case <-worker.stop:
+			l4g.Debug("Worker %v: Received stop signal", worker.name)
+			running = false
+		case job := <-worker.jobs:
+			l4g.Debug("Worker %v: Received a new candidate job.", worker.name)
+			worker.DoJob(&job)
+		}
+	}
+
+	l4g.Debug("Worker %v: Finished", worker.name)
+	worker.stopped <- true
+}
+
+func (worker *TestWorker) DoJob(job *model.Job) {
+	if claimed, err := ClaimJob(worker.store, job); err != nil {
+		l4g.Error("Job: %v: Error occurred while trying to claim job: %v", job.Id, err.Error())
+		return
+	} else if !claimed {
+		return
+	}
+
+	cancelCtx, cancelCancelWatcher := context.WithCancel(context.Background())
+	cancelWatcherChan := make(chan interface{}, 1)
+	go CancellationWatcher(cancelCtx, job.Id, worker.store, cancelWatcherChan)
```

Comment:
Yeah, I like that. To be honest, it's probably nicer to replace all the `for running {}` bits with a deferred cleanup function.